lib/ansible/plugins/filter/core.py

```diff
@@ -299,25 +299,36 @@ def mandatory(a, msg=None):
 
 
 def combine(*terms, **kwargs):
-    recursive = kwargs.get('recursive', False)
-    if len(kwargs) > 1 or (len(kwargs) == 1 and 'recursive' not in kwargs):
-        raise AnsibleFilterError("'recursive' is the only valid keyword argument")
-
-    dicts = []
-    for t in terms:
-        if isinstance(t, MutableMapping):
-            recursive_check_defined(t)
-            dicts.append(t)
-        elif isinstance(t, list):
-            recursive_check_defined(t)
-            dicts.append(combine(*t, **kwargs))
-        else:
-            raise AnsibleFilterError("|combine expects dictionaries, got " + repr(t))
-
-    if recursive:
-        return reduce(merge_hash, dicts)
-    else:
-        return dict(itertools.chain(*map(iteritems, dicts)))
+    recursive = kwargs.pop('recursive', False)
+    list_merge = kwargs.pop('list_merge', 'replace')
+    if kwargs:
+        raise AnsibleFilterError("'recursive' and 'list_merge' are the only valid keyword arguments")
+
+    # allow the user to do `[dict1, dict2, ...] | combine`
+    dictionaries = flatten(terms, levels=1)
```

Comment:
Why not leave `levels=1` away? From your PR description, this breaks backwards compatibility for no reason (or if there's a reason, it's hidden somewhere).

Reply:
> non backward compatible: `[dict1, [dict2]] | combine` doesn't work anymore, because I thought it shouldn't. If you disagree, `flatten(terms, levels=1)` → `flatten(terms)` and it's fixed.

There is no technical reason.
Without this PR, `[dict1, [dict2]] | combine` is working but it's an undocumented behavior. As an ansible user, when I provide bad input to a filter I expect it to raise an error instead of guessing what I want to achieve.
Providing `[dict1, [dict2]]` to `combine` is a bad input as nothing in the documentation state it will flatten the list.

Though it's up for debate and if you disagree I will gradely remove the `levels=1`. I don't want that to block the entire PR.

@felixfontein What do you think ?

Reply:
If this is a bug that's fixed, there should be a changelog fragment (with `bugfixes:`) that mentions that this was a bug that is now fixed. I am fine if this is fixed / the undocumented behavior removed, but it should be clearly documented.

(If someone else complains, maybe move it into a separate PR, resp. a follow-up PR. If nobody else complains, just leave it there - assuming there is a bugfix changelog fragment for thix.)

Reply:
Ok, thanks !
I [created a changelog fragment](https://github.com/ansible/ansible/pull/57894/commits/3c323c1b0c1d803794620a57105d4b3482ba74b4):
```
+bugfixes:
+- combine filter - ``[dict1, [dict2]] | combine`` now raise an error; previously ``combine`` had an undocumented behaviour where it was flattening the list before combining it (https://github.com/ansible/ansible/pull/57894#discussion_r339517518).
+minor_changes:
+- combine filter - now accept a ``list_merge`` argument which modifies its behaviour when the hashes to merge contain arrays/lists.
```